controller/volume_controller.go

```diff
@@ -383,6 +383,25 @@ func (vc *VolumeController) ReconcileEngineReplicaState(v *longhorn.Volume, e *l
 	}
 
 	if e.Status.CurrentState != types.InstanceStateRunning {
+		// Corner case: when vc found healthyCount is 0, vc failed to update volume to faulted state
+		// and volume may be detached (engine stopped state).
+		// Then the volume needs to be reconciled to the faulted state here.
+		failedCount := 0
+		for _, r := range rs {
+			if r.Spec.FailedAt != "" {
+				failedCount++
+			}
+			if failedCount >= v.Spec.NumberOfReplicas {
```

Comment:
It's wrong.

A volume can have 10 failed replicas, but as long as there is one good replica, it's not faulted.

The condition should be no healthy replicas. `FailedAt == "" && HealthyAt != ""`

Reply:
It seems that `HealthyAt != ""` is unnecessary. Here the volume is not running and `HealthyAt` can be empty.

Reply:
Right. The one case may result in `FailedAt == "" && HealthyAt == ""` should be cleaned up by https://github.com/longhorn/longhorn-manager/blob/master/controller/volume_controller.go#L498